File: bab_functions.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <mpi.h>
#include "biqbin.h"
  
  /* defined in heap.c */  
extern Heap *heap;  
extern int main_problem_size;           
extern BabSolution *solution;
   
extern BiqBinParameters params;
extern GlobalVariables globals;

extern int num_workers_used;

/* initialize global lower bound to 0 and global solution vector to zero */
void initializeBabSolution() { 

    BabSolution bs;

    for (int i = 0; i < main_problem_size; ++i) {
        bs.X[i] = 0;
    }

    init_solution_lb(0, &bs);
}

/* NOTE: int *sol in functions evaluateSolution and updateSolution have length main_problem_size
 * -> to get objecive multiple with Laplacian that is stored in upper left corner of SP->L
 */
double evaluateSolution(const int *sol, const Problem *SP) {

    double val = 0.0;
    int problem_size = SP->n - 1;
    for (int i = 0; i < problem_size; ++i) {
        for (int j = 0; j < problem_size; ++j) {
            val += SP->L[j + i * SP->n] * sol[i] * sol[j];
        }
    }

    return val;
}


/*
 * Only this function can update best solution and value.
 * Returns 1 if success.
 */
int updateSolution(const int *x, const Problem *SP) {
    
    int solutionAdded = 0;
    double sol_value;
    BabSolution solx;

    // Copy x into solx --> because update_lower_bound needs BabSolution and not int*
    for (int i = 0; i < SP->n - 1; ++i) {
      solx.X[i] = x[i];
    }

    sol_value = evaluateSolution(x, SP); // computes objective value of solx

    /* If new solution is better than the global solution,
     * then update and print the new solution. */
    if (update_lower_bound(sol_value, &solx)) {
        solutionAdded = 1;
    }
    
    return solutionAdded;
}


/* MASTER process main routine */
void master_Bab_Main(Message message, int source, int *busyWorkers, int numbWorkers, int *numbFreeWorkers, MPI_Datatype BabSolutiontype) {

    // If the algorithm stops before finding the optimal solution
    if (!globals.stopped && (params.time_limit > 0 && (MPI_Wtime() - globals.TIME) > params.time_limit) ) {
        
        // signal to printFinalOutput that algorihtm stopped early
        globals.stopped = 1;
    }

    MPI_Status status;

    switch(message) {

        case IDLE:

            busyWorkers[source] = 0;
            ++(*numbFreeWorkers);
            break;  

        case NEW_VALUE:
        {
        
            // receive best lower bound and corresponding feasible solution
            double g_lowerBound;
            BabSolution solx;
            
            MPI_Recv(&g_lowerBound, 1, MPI_DOUBLE, source, LOWER_BOUND, MPI_COMM_WORLD, &status);
            MPI_Recv(&solx, 1, BabSolutiontype, source, SOLUTION, MPI_COMM_WORLD, &status);  

            if ( update_lower_bound(g_lowerBound, &solx) ){
                printf("Feasible solution %.0lf\n", get_lower_bound());
            }
            
            // send update information back to worker
            g_lowerBound = get_lower_bound();

            MPI_Send(&g_lowerBound, 1, MPI_DOUBLE, source, LOWER_BOUND, MPI_COMM_WORLD);
            break;       
        }
        case SEND_FREEWORKERS:
        {            
            // get number of requested workers            
            int workers_request;                
            MPI_Recv(&workers_request, 1, MPI_INT, source, FREEWORKER, MPI_COMM_WORLD, &status);
                        
            // compute number of freeworkers
            int num_workers_available = (workers_request < *numbFreeWorkers) ? workers_request : *numbFreeWorkers;
            int available_workers[num_workers_available];

            for(int i = 1, j = 0; (i < numbWorkers) && (j < num_workers_available); ++i)    // master has rank 0 and is not considered
            {
                if(busyWorkers[i] == 0){ // is free
                    available_workers[j] = i;
                    ++j;
                    busyWorkers[i] = 1; // set to busy
                    --(*numbFreeWorkers);
                }
            }
    

            // worker branched subproblem in local queue --> add 2 bab nodes
            increase_num_eval_nodes();
            increase_num_eval_nodes(); 

	        // count current number of busy workers
	        int current_busy = 0;

	        for (int i = 1; i < numbWorkers; ++i) {
		       if (busyWorkers[i] == 1)
		       ++current_busy;
	        } 
	
	        num_workers_used = (current_busy > num_workers_used) ? current_busy : num_workers_used;

	        // send message back
            double g_lowerBound = get_lower_bound();            
            MPI_Send(&num_workers_available, 1, MPI_INT, source, NUM_FREE_WORKERS, MPI_COMM_WORLD);              
            MPI_Send(available_workers, num_workers_available, MPI_INT, source, FREEWORKER, MPI_COMM_WORLD);
            MPI_Send(&g_lowerBound, 1, MPI_DOUBLE, source, LOWER_BOUND, MPI_COMM_WORLD);     
            break;
        }
    }
}

/* print solution 0-1 vector */
void printSolution(FILE *file) {

    fprintf(file, "Solution = ( ");
    for (int i = 0; i < main_problem_size; ++i) {
        if (solution->X[i] == 1) {
            fprintf(file, "%d ", i + 1);
        }
    }
    fprintf(file, ")\n");
}


/* print final output */
void printFinalOutput(FILE *file, int num_nodes) {

    // Best solution found
    double best_sol = get_lower_bound();

    fprintf(file, "\nNodes = %d\n", num_nodes);
    
    // normal termination
    if (!globals.stopped) {
        fprintf(file, "Root node bound = %.2lf\n", globals.root_bound);
        fprintf(file, "Maximum value = %.0lf\n", best_sol);
        
    } else { // B&B stopped early
        fprintf(file, "TIME LIMIT REACHED.\n");
        fprintf(file, "Root node bound = %.2lf\n", globals.root_bound); 
        fprintf(file, "Best value = %.0lf\n", best_sol);
    }

    printSolution(file);
    fprintf(file, "Time = %.2f s\n\n", MPI_Wtime() - globals.TIME);
}


/* Bab function called at the end of the execution.
 * This function frees the memory allocated by the program. */
void Bab_End(void) {
    freeMemory(&globals);   
}


/*
 * getBranchingVariable function used in the Bab_GenChild routine to determine
 * which variable x[ic] to branch on.
 *
 * node: the current node of the branch-and-bound search tree
 */
int getBranchingVariable(BabNode *node) {

    int ic = -1;  // x[ic] is the variable to branch on
    double maxValue, minValue;

    /* 
     * Choose the branching variable x[ic] based on params.branchingStrategy
     */
    if (params.branchingStrategy == LEAST_FRACTIONAL) {
        // Branch on the variable x[ic] that has the least fractional value
        maxValue = -BIG_NUMBER;
        for (int i = 0; i < main_problem_size; ++i) {
            if (!(node->xfixed[i]) && fabs(0.5 - node->fracsol[i]) > maxValue) {
                ic = i;
                maxValue = fabs(0.5 - node->fracsol[ic]);
            }
        }
    }
    else if (params.branchingStrategy == MOST_FRACTIONAL) {
        // Branch on the variable x[ic] that has the most fractional value
        minValue = BIG_NUMBER;
        for (int i = 0; i < main_problem_size; ++i) {
            if (!(node->xfixed[i]) && fabs(0.5 - node->fracsol[i]) < minValue) {
                ic = i;
                minValue = fabs(0.5 - node->fracsol[ic]);
            }
        }
    }
    else {
        fprintf(stderr, "Error: Wrong value for params.branchingStrategy\n");
        MPI_Abort(MPI_COMM_WORLD,10);
    }

    return ic;
}


/* Count the number of fixed variables */
int countFixedVariables(BabNode *node) {
    
    int numFixedVariables = 0;

    for (int i = 0; i < main_problem_size; ++i) {
        if (node->xfixed[i]) {
            ++numFixedVariables;
        }
    }

    return numFixedVariables;
}

File: bab_functions.py
import heapq
from typing import Optional
import numpy as np
from biqbin_data_objects import BiqBinParameters, _BabSolution, _BabNode
# Branch and bound helper functions


class BabFunctions:
    def __init__(self, L_mat, num_vert: int, params: BiqBinParameters):
        self.L_mat = L_mat
        self.problem_size = num_vert - 1
        self.root_bound = None

        self.branching_strategy = params.branchingStrategy

        self.solution: Optional[_BabSolution] = None
        self.best_lower_bound: Optional[float] = 0

        self.num_eval_nodes = 1

        self.pq = []

    def generate_node(self, parent_node: Optional[_BabNode] = None):
        node = _BabNode()

        for i in range(self.problem_size):  # Loop over all possible indices
            if parent_node is None:
                node.xfixed[i] = 0
                node.sol.X[i] = 0
            else:
                node.xfixed[i] = parent_node.xfixed[i]
                node.sol.X[i] = parent_node.sol.X[i] if node.xfixed[i] else 0

        # Set level: root node starts at 0, children are +1 from parent
        node.level = 0 if parent_node is None else parent_node.level + 1
        return node

    # Convert BabNode c_type object to numpy array for MPI transit
    def babnode_to_numpy(self, node):
        arr = np.zeros(3 * self.problem_size + 2, dtype=np.float64)
        arr[:self.problem_size] = np.frombuffer(
            node.xfixed, dtype=np.int32, count=self.problem_size).astype(np.float64)
        arr[self.problem_size:2 * self.problem_size] = np.frombuffer(
            node.sol.X, dtype=np.int32, count=self.problem_size).astype(np.float64)
        arr[2*self.problem_size:3 * self.problem_size] = np.frombuffer(
            node.fracsol, dtype=np.float64, count=self.problem_size)
        arr[3*self.problem_size] = float(node.level)
        arr[3*self.problem_size+1] = node.upper_bound
        return arr

    # Convert the numpy array back into BabNode to be able to pass it into the solver
    def numpy_to_babnode(self, arr):
        node = _BabNode()
        # Only update first NMAX elements; leave rest of the full-sized buffer untouched
        node.xfixed[:self.problem_size] = arr[:self.problem_size].astype(
            np.int32)
        node.sol.X[:self.problem_size] = arr[self.problem_size:2 *
                                             self.problem_size].astype(np.int32)
        node.fracsol[:self.problem_size] = arr[2 *
                                               self.problem_size:3*self.problem_size]

        node.level = int(arr[3*self.problem_size])
        node.upper_bound = arr[3*self.problem_size + 1]

        return node

    def get_branching_variable(self, node: _BabNode):
        # Convert ctypes arrays to NumPy arrays
        xfixed = np.frombuffer(node.xfixed, dtype=np.int32)[:self.problem_size]
        fracsol = np.frombuffer(node.fracsol, dtype=np.float64)[
            :self.problem_size]

        # Create a boolean mask for unfixed variables (True = not fixed, False = fixed)
        unfixed_mask = xfixed == 0
        fractional_diffs = np.abs(0.5 - fracsol)

        if self.branching_strategy == 0:  # Least fractional
            # Mask values for fixed variables
            ic = np.argmax(fractional_diffs * unfixed_mask)
        elif self.branching_strategy == 1:  # Most fractional
            # Use np.where() safely
            ic = np.argmin(np.where(unfixed_mask, fractional_diffs, np.inf))
        else:
            raise ValueError("Error: Wrong value for branching_strategy")

        # Ensure only valid indices are returned
        return ic if unfixed_mask[ic] else -1

    def count_fixed_variables(self, node: _BabNode) -> int:
        """
        Counts the number of fixed variables in the BabNode.

        Parameters:
        - node: BabNode object

        Returns:
        - The number of fixed variables
        """
        xfixed = np.frombuffer(node.xfixed, dtype=np.int32)[:self.problem_size]
        return np.count_nonzero(xfixed)

    def is_leaf_node(self, node: _BabNode) -> bool:
        return self.count_fixed_variables(node) == self.problem_size

    # branches of a node, calls evaluate for both children nodes using C biqbin,
    # if needed pushes them into the priority queue
    def branch(self, node: _BabNode, biqbin, rank: int):
        biqbin.evaluate(node, rank)

        if (self.best_lower_bound + 1 > node.upper_bound):
            return

        ic = self.get_branching_variable(node)

        for xic in range(2):  # 0 - 1
            child_node = self.generate_node(node)
            child_node.xfixed[ic] = 1
            child_node.sol.X[ic] = xic
            self.num_eval_nodes += 1
            if (self.is_leaf_node(child_node)):
                sol_val = self.evaluate_solution(child_node.sol)
                if (sol_val > self.best_lower_bound):
                    self.solution = child_node.sol
                    self.best_lower_bound = sol_val
                    print(
                        f"Node {self.num_eval_nodes} Feasible solution {sol_val:2}")
            else:
                # biqbin.evaluate(child_node, rank)
                heapq.heappush(self.pq, (-child_node.upper_bound, child_node))

    def evaluate_solution(self, sol: _BabSolution) -> float:
        sol_val = 0
        for i in range(self.problem_size):
            for j in range(self.problem_size):
                sol_val += self.L_mat[i][j] * sol.X[i] * sol.X[j]
        return sol_val

    def get_solution_string(self):
        sol_str = "Solution = ("
        for i in range(self.problem_size):
            if self.solution.X[i] == 1:
                sol_str += f" {i + 1}"
        sol_str += " )"
        return sol_str
